File: solutions/c/luhn/2/luhn.c
#include "luhn.h"
#include <stddef.h>
#include <ctype.h>
#include <stdbool.h>
#include <string.h>

bool luhn(const char* digits) {
    size_t digit_count = 0;

    for (size_t j = 0; digits[j]; j++) {
        if (isdigit(digits[j])) {
            digit_count++;
            continue;
        }
        if (digits[j] == ' ') {
            continue;
        }
        return false;
    }
    if (digit_count < 2) return false;

    size_t i = strlen(digits);

    int checksum = 0;
    bool should_double = false;

    for (size_t index = i; index-- > 0; ) {
        if (digits[index] == ' ') continue;

        int value = digits[index] - '0';

        if (should_double) {
            value *= 2;
            if (value > 9) value -= 9;
        }

        checksum += value;
        should_double = !should_double;
    }

    return checksum % 10 == 0;
}

File: solutions/elixir/luhn/2/lib/luhn.ex
defmodule Luhn do

  @spec valid?(String.t()) :: boolean()
  def valid?(digits),
    do: with digits <- normalize(digits),
    do: validate(digits)

  @spec normalize(String.t()) :: charlist()
  defp normalize(digits),
    do: String.to_charlist(String.replace(digits, " ", ""))

  @spec validate(charlist()) :: boolean()
  defp validate(digits) when length(digits) <= 1, do: false
  defp validate(digits) do
    Enum.all?(digits, &(&1 >= ?0 and &1 <= ?9)) and
      rem(checksum(digits), 10) == 0
  end

  @spec checksum(charlist()) :: integer()
  defp checksum(digits) do
    digits
    |> Enum.reverse()
    |> Enum.with_index()
    |> Enum.map(&adjust/1)
    |> Enum.sum()
  end

  @spec adjust({integer(), integer()}) :: integer()
  defp adjust({digit, index}) when
    rem(index, 2) == 1, do: adjust(digit)
  defp adjust({digit, _index}), do: digit - ?0

  @spec adjust(integer()) :: integer()
  defp adjust(digit) do
    doubled = (digit - ?0) * 2
    if doubled > 9,
      do:   doubled - 9,
      else: doubled
  end

end
